lib/spack/spack/package.py

```diff
@@ -956,7 +956,19 @@ def provides(self, vpkg_name):
 
     @property
     def installed(self):
-        return os.path.isdir(self.prefix)
+
+        has_prefix = os.path.isdir(self.prefix)
+        try:
+            # If the spec is in the DB, check the installed
+            # attribute of the record
+            rec = spack.store.db.get_record(self.spec)
+            db_says_installed = rec.installed
```

Comment:
@scheibelp @tgamblin I was wondering what is a case that could leave us with a record in the DB that has `rec.installed ==False`. Do you have any in mind?

Reply:
When packages are uninstalled the corresponding record is not removed - instead it is marked as `.installed = False`

Reply:
It doesn't seem the case to me, that's why I was wondering if I missed some corner case where it could happen:
```console
$ spack find
==> 0 installed packages.

$ cat opt/spack/.spack-db/index.json 
{
 "database": {
  "installs": {},
  "version": "0.9.3"
 }
}
$ spack install zlib
==> Installing zlib
==> Fetching file:///home/mculpo/production/mirror/zlib/zlib-1.2.11.tar.gz
...
  Fetch: 0.02s.  Build: 4.33s.  Total: 4.35s.
[+] /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-8/zlib-1.2.11-ypfhlhjehavum42gl6uvax7o6kzhseoi

$ cat opt/spack/.spack-db/index.json 
{
 "database": {
  "installs": {
   "ypfhlhjehavum42gl6uvax7o6kzhseoi": {
    "explicit": true,
    "installation_time": 1526021395.450302,
    "ref_count": 0,
    "installed": true,
    "path": "/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-8/zlib-1.2.11-ypfhlhjehavum42gl6uvax7o6kzhseoi",
    "spec": {
     "zlib": {
     ...
}

$ spack uninstall -y zlib
==> Successfully uninstalled zlib@1.2.11%gcc@8+optimize+pic+shared arch=linux-ubuntu18.04-x86_64 /ypfhlhj

$ cat opt/spack/.spack-db/index.json 
{
 "database": {
  "installs": {},
  "version": "0.9.3"
 }
}
```

Reply:
I'm not sure how that comes about, nevertheless it is safest to proceed by checking the `.installed` field of the DB record since that is there.

@becker33 do you have a notion of when the `.installed` field on a DB record is used?